### PR TITLE
add remove-unchanged API

### PR DESCRIPTION
**Pull request by humorless**
Try to solve the issue #13 with an API `remove-unchanged`, test, CHANGLOG

### CODE REVIEW

File: src/lambdaisland/deep_diff2/strip.cljc [hunk 1]
```diff
@@ -0,0 +1,43 @@
+(ns lambdaisland.deep-diff2.strip
```
Comment (alysbrooks) [inline src/lambdaisland/deep_diff2/strip.cljc:1]: I think this should be called `lambdaisland.deep-diff2.manipulate`, or something else that's broader than "strip."

File: src/lambdaisland/deep_diff2/strip.cljc [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      (instance? Insertion x)))
+
+(defn extend-flatten
+  "Flatten, which can apply on hashmap"
+  [x]
+  (filter (complement coll?)
+          (rest (tree-seq coll? seq x))))
+
+(defn has-diff-item?
+  "Checks if there are any diff items in x or sub-tree of x"
```
Comment (alysbrooks) [inline src/lambdaisland/deep_diff2/strip.cljc:22]: Maybe use `coll` instead of `x`?

File: src/lambdaisland/deep_diff2/strip.cljc [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+(defn diff-item?
+  "Checks if x is a Mismatch, Deletion, or Insertion"
+  [x]
+  (or (instance? Mismatch x)
+      (instance? Deletion x)
+      (instance? Insertion x)))
+
+(defn extend-flatten
+  "Flatten, which can apply on hashmap"
+  [x]
```
Comment (alysbrooks) [inline src/lambdaisland/deep_diff2/strip.cljc:17]: Maybe use `coll` instead of `x`?

File: src/lambdaisland/deep_diff2/strip.cljc [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  (postwalk
+   (fn [x]
+     (cond
+       (map-entry? x) (cond
+                        (diff-item? (key x)) (do
+                                               ;(println "cond 1, keep" x)
+                                               x)
+                        (has-diff-item? (val x)) (do
+                                                    ;(println "cond 2, keep" x)
+                                                   x))
```
Comment (alysbrooks) [inline src/lambdaisland/deep_diff2/strip.cljc:40]: Could this just be something like `(while (or  (diff-item? (key x)) (has-diff-item? (val x))) x)`?

**Review by alysbrooks — COMMENTED** [review]
Looks good, but there's some cleanup needed, I think.

If you wanted to add some property-based/generative testing, you could add a test that verifies that, for any data structure, diffing it against itself and then applying `remove-unchanged` returns an empty structure.

File: src/lambdaisland/deep_diff2/strip.cljc (hunk 2 above)
Comment (humorless) [reply src/lambdaisland/deep_diff2/strip.cljc:22]: The `has-diff-item?` applies to the val item of a `map-entry`. It could be any type, so I choose `x`.

File: src/lambdaisland/deep_diff2/strip.cljc (hunk 3 above)
Comment (humorless) [reply src/lambdaisland/deep_diff2/strip.cljc:17]: I just follow most of the details [here](https://github.com/clojure/clojure/blob/clojure-1.10.1/src/clj/clojure/core.clj#L7136).

File: src/lambdaisland/deep_diff2/strip.cljc (hunk 4 above)
Comment (humorless) [reply src/lambdaisland/deep_diff2/strip.cljc:40]: yes. I will do that.

**Comment by humorless** [comment]
Thanks @alysbrooks  so much. 
After I tried to run the property test, I found out that there are a lot of edge cases which I did not think it throughly. 

I can not make this property test pass until now.
However, I have considered more about the semantic of `remove-unchanged` and I changed it a lot. 

```
(defspec diff-itself 100
  (prop/for-all [x diff-test/gen-any-except-NaN]
                (if (coll? x)
                  (= (manipulate/remove-unchanged (ddiff/diff x x))
                     (empty x))
                  (= (manipulate/remove-unchanged (ddiff/diff x x))
                     nil))))
```

@plexus  @alysbrooks 
Do you think that I should dig it throughly to figure out why this property test can not work?  Or the current `remove-unchanged` you think is good enough?

**Comment by humorless** [comment]
> What do the failing cases look like?

```
FAIL in cljs:lambdaisland.deep-diff2.manipulate-test/diff-itself (cljs/test.js:429)
Expected:
  {:result true}
Actual:
  {:shrunk
   {:total-nodes-visited 756,
    :depth 51,
    :pass? false,
    :result false,
    :result-data nil,
    :time-shrinking-ms 168,
    :smallest [#{0 1 false -6 -5 -4 -3 -2 -1}]},
   :failed-after-ms 70,
   :num-tests 59,
   :seed 1680177300377,
   :fail
   [#{1.1197369622161879e-14 1.3019822841584656e-21 0.6875
      #uuid "a907a7fe-d2eb-482d-b1cc-3acfc12daf55"
      -30
      :X/*!1:3
      :u7*A/p?2IG5d*!Nl
      :**d7ws
      "ý"
      "ÔB*àñS¬ÚûV¡ç¯±·á£H
                            û?'V$ëY;CLk-oOV"
      !U-h_C*A7/x0_n1
      A-*wn./o_?4w18-!
      "ìêÜ¼à4^¤mÐðktê1_ò· À4\n@J\"29)cd-\t®"
      y3W-2
      #uuid "6d507164-f8b9-401d-8c44-d6b0e310c248"
      "M"
      :cy7-3
      :w4/R.-s?9V5
      #uuid "1bcb00c9-88b9-4eae-9fea-60600dfaefa0"
      -20
      #uuid "269ab6f9-f19d-4c9d-a0cb-51150e52e9f7"
      -235024979
      :O:m_9.9+A/N+usPa6.HA*G
      228944.657438457
      :x/w?
      :__+o+sut9!t/?0l
      "â«"
      false
      #uuid "b6295f83-8176-47b5-946e-466f74226629"
      e3zQ!E*5
      :T5rb
      :++y:2
      -7364
      zG/ex23
      "¡"
      -4318364480
      :D+?2?!/Hrc!jA7z_2
      :z-I/!8Uq+d?
      -0.5588235294117647
      -0.5925925925925926
      -0.8108108108108109}],
   :result false,
   :result-data nil,
   :failing-size 58, 
   :pass? false, 
   :test-var "diff-itself"}
```

**Comment by plexus** [comment]
The problem is not in `remove-unchanged`, but in deep-diff itself. It seems for some sets we do some weird things:

```clj
(let [s #{false 5}]
  (ddiff/diff s s))
;; => #{{:- 5} false {:+ 5}}
```

**Comment by alysbrooks** [comment]
A little surprised that bug hasn't been discovered before. Maybe it's just sets with a mixture of types that have issues?

**Comment by plexus** [comment]
Two remarks, 
- you have a `diff-item?` where you check with `instance?` but then later on you were checking for `:-` or `:+`, isn't that a problem if the user passes in a map with `:+` or `:-` key?
- you were checking for `diff-item?` or map keys, but map keys can also be arbitrary values, so we need to also check that none of their children have diff items.

Actually, now that I write this I'm not sure if we descend into map keys when diffing... 

Anyway, great work, happy to have this feature. Next we can think about how to expose this in kaocha.

**Comment by humorless** [comment]
The two remarks are totally true. I only thought of the most typical cases when implemented.
